packages/react-form-state/src/utilities.ts

```diff
@@ -1,3 +1,61 @@
+export function deepEqual(first: any, second: any, maxDepth = 2000, depth = 0) {
```

Comment:
@TheMallen did you just come up with this code, or did you copy it from somewhere?

Reply:
Came up with it, though I've seen many versions of this written before.

Reply:
fair enough. I just wonder what the nuanced differences are in behavior here. Also, what is the performance tradeoff over using `lodash.isequal`? Is it worth the risk of deviating from an implementation that is industry-trusted and battle tested? Perhaps we will encounter bugs with this implementation that are unexpected?

Reply:
I don't think runtime microperformance is the right thing to optimize for. We should worry about bundle size, since it's already a problem for us, and causing too many rerenders (which this does not effect), and only worry about microperf if we encounter a real issue. That being said, even if they had some black magic going on to speed up deep equality operations that makes it ten times faster I don't think that would actually be worth the weight. The only two places we use this are comparing new initialValues coming in and comparing fields to their initial values on update. the former will very seldomly happen (only after a save or on first mount), and the latter is 90 percent of the time going to be on scalar values. 

As for edgecases we have to remember that a generic library version of something like this needs to worry about a much wider variety of objects than we do. We should not be worrying about things like objects with weird crazy prototype setups, or even just ones with getters and setters. In fact we shouldn't even care about objects with methods. The only things these are being applied to are form data which *should always* be a plain old javascript object.

Reply:
It also might actually be a good idea to drop `deepEquals/isEqual` from the updater check entirely and just use `Object.is` there, since we expect immutable updates to those anyway.